src/app/CommandHandler.cpp

```diff
@@ -97,6 +98,60 @@ void CommandHandler::OnInvokeCommandRequest(Messaging::ExchangeContext * ec, con
     mGoneAsync = true;
 }
 
+CHIP_ERROR CommandHandler::ValidateCommands(TLV::TLVReader & invokeRequestsReader)
+{
+    CHIP_ERROR err          = CHIP_NO_ERROR;
+    size_t commandCount     = 0;
+    bool commandRefExpected = false;
+
+    TLV::Utilities::Count(invokeRequestsReader, commandCount, false /* recurse */);
+    if (IsGroupRequest())
+    {
+        VerifyOrReturnError(commandCount == 1, CHIP_ERROR_INVALID_ARGUMENT);
+        return CHIP_NO_ERROR;
+    }
+
+    VerifyOrReturnError(commandCount <= CHIP_CONFIG_MAX_PATHS_PER_INVOKE, CHIP_ERROR_INVALID_ARGUMENT);
+    commandRefExpected = commandCount > 1;
+
+    while (CHIP_NO_ERROR == (err = invokeRequestsReader.Next()))
+    {
+        VerifyOrReturnError(TLV::AnonymousTag() == invokeRequestsReader.GetTag(), CHIP_ERROR_INVALID_ARGUMENT);
+        CommandDataIB::Parser commandData;
+        ReturnErrorOnFailure(commandData.Init(invokeRequestsReader));
+
+        CommandPathIB::Parser commandPath;
+        ConcreteCommandPath concretePath(0, 0, 0);
+
+        ReturnErrorOnFailure(commandData.GetPath(&commandPath));
+        ReturnErrorOnFailure(commandPath.GetConcreteCommandPath(concretePath));
+
+        Optional<uint16_t> commandRef;
+        uint16_t ref;
+        err = commandData.GetRef(&ref);
+        VerifyOrReturnError(err == CHIP_NO_ERROR || err == CHIP_END_OF_TLV, err);
+        if (err == CHIP_END_OF_TLV && commandRefExpected)
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+        if (err == CHIP_NO_ERROR)
+        {
+            commandRef = MakeOptional(ref);
+        }
+
+        VerifyOrReturnError(mCommandRefLookupTable.IsRequestedPathAndRefUnique(concretePath, commandRef),
+                            CHIP_ERROR_INVALID_ARGUMENT);
```

Comment:
Is CHIP_ERROR_INVALID_ARGUMENT unique enough to indicate failure of validation? In a sense, it sounds like every error would cause ProcessInvokeRequest to end-up failing.

Reply:
These are all invalid arguments to invoke the command. That and the fact that `ProcessInvokeRequest` sends a `Status::InvalidAction` for any/all of these validation errors in spec right now/ I guess we could change spec, but there really isn't any granularity gain from the spec IM Status